Foundation/NSDecimal.swift

```diff
@@ -1,155 +1,1838 @@
+            } else {
+                return .overflow
+            }
+        }
+    }
+
+    while big._length != 0 && big[big._length - 1] == 0 {
+        big._length -= 1
+    }
+
+    return .noError
+}
+
+fileprivate func integerDivide(_ r: inout WideDecimal,
+                               _ cu: WideDecimal,
+                               _ cv: WideDecimal) -> NSDecimalNumber.CalculationError {
+    // Calculate result = a / b.
+    // Result could NOT be a pointer to same space as a or b.
+    // resultLen must be >= aLen - bLen.
+    //
+    // Based on algorithm in The Art of Computer Programming, Volume 2,
+    // Seminumerical Algorithms by Donald E. Knuth, 2nd Edition. In addition
+    // you need to consult the erratas for the book available at:
+    //
+    //   http://www-cs-faculty.stanford.edu/~uno/taocp.html
+
+    var u = WideDecimal(true)
+    var v = WideDecimal(true) // divisor
+
+    var v1:UInt16, v2:UInt16
+
+    // Simple case
+    if cv.isZero {
+        return .divideByZero;
+    }
+
+    // If u < v, the result is approximately 0...
+    if cu < cv {
+        r._length = 0
+        return .noError;
+    }
+
+    // Fast algorithm
+    if cv._length == 1 {
+        r = cu
+        let (_,error) = divideByShort(&r, cv[0])
+        return error
+    }
+
+    u._length = cu._length + 1
+    v._length = cv._length + 1
+
+    // D1: Normalize
+    // Calculate d such that d*highest_digit of v >= b/2 (0x8000)
+    //
+    // I could probably use something smarter to get d to be a power of 2.
+    // In this case the multiply below became only a shift.
+    let d: UInt32 = UInt32((1 << 16) / Int(cv[cv._length - 1] + 1))
+
+    // This is to make the whole algorithm work and u*d/v*d == u/v
+    _ = multiplyByShort(&u, UInt16(d))
+    _ = multiplyByShort(&v, UInt16(d))
+
+    // Set a zero at the leftmost u position if the multiplication
+    // do not have a carry.
+    if(u._length == cu._length) {
+        u[u._length] = 0
+        u._length += 1
+    }
+
+    v[v._length] = 0; // Set a zero at the leftmost v position.
+    // the algorithm will use it during the
+    // multiplication/subtraction phase.
+
+    // Determine the size of the quotient.
+    // It's an approximate value.
+    let ql:UInt16 = UInt16(u._length - v._length)
+
+    // Some useful constants for the loop
+    // It's used to determine the quotient digit as fast as possible
+    // The test vl > 1 is probably useless, since optimizations
+    // up there are taking over this case. I'll keep it, just in case.
+    v1 = v[v._length-1]
+    v2 = v._length > 1 ? v[v._length-2] : 0
+
+    // D2: initialize j
+    // On each pass, build a single value for the quotient.
+    for j in 0..<ql {
+        var q: UInt32 // Quotient digit. could be a short.
+
+        // D3: calculate q^
+        // This formula and test for q gives at most q+1; See Knuth for proof.
+
+        let ul = u._length
+        let uu = u // work around compiler bug
```

Comment:
might be good to attach bugs.swift.org url here for a filed ticket just so it is known what bug it is